url64d634b8f0_agdturner_generic_tgz-pJ8-uk_ac_leeds_ccg_andyt_generic_visualisation_charts_Generic_BarChartJ8/src/uk/ac/leeds/ccg/andyt/generic/math/Generic_BigDecimal.java

```diff
@@ -0,0 +1,7278 @@
+            decimalPlaces = probabilityScale;
+        }
+        BigDecimal midTestValue = new BigDecimal("0.5");
+        BigDecimal half_BigDecimal = new BigDecimal("0.5");
+        BigDecimal two_BigDecimal = new BigDecimal("2.0");
+        if (probability.compareTo(midTestValue) == 0) {
+            return a_Random.nextBoolean();
+        }
+        if (probability.compareTo(midTestValue) == -1) {
+            return randomTest(
+                    a_Random,
+                    probability,
+                    BigDecimal.ZERO,
+                    BigDecimal.ONE,
+                    midTestValue,
+                    half_BigDecimal,
+                    two_BigDecimal,
+                    decimalPlaces,
+                    a_RoundingMode);
+        } else {
+            return !randomTest(
+                    a_Random,
+                    BigDecimal.ONE.subtract(probability),
+                    BigDecimal.ZERO,
+                    BigDecimal.ONE,
+                    midTestValue,
+                    half_BigDecimal,
+                    two_BigDecimal,
+                    decimalPlaces,
+                    a_RoundingMode);
+        }
+    }
+
+    /**
+     * Effectively this is the same as generating a random number between 0 and
+     * 1 and comparing it with probability and if it were higher then return
+     * false and otherwise return true
+     *
+     * @param a_Random
+     * @param probability 0 <= probability <=1
+     * @param a_MathContext
+     * @return true or false based on a random uniform test of probability
+     */
+    public static boolean randomUniformTest(
+            Random a_Random,
+            BigDecimal probability,
+            MathContext a_MathContext) {
+        // Special case probability <= 0
+        if (probability.compareTo(BigDecimal.ZERO) != 1) {
+            if (probability.compareTo(BigDecimal.ZERO) == 0) {
+                return false;
+            } else {
+                System.out.println(
+                        "Warning probabilty negative in "
+                        + Generic_BigDecimal.class
+                        + ".randomUniformTest(Random,BigDecimal). "
+                        + "Returning false.");
+                return false;
+            }
+        }
+        // Special case probability >= 1
+        if (probability.compareTo(BigDecimal.ONE) != -1) {
+            if (probability.compareTo(BigDecimal.ONE) == 0) {
+                return true;
+            } else {
+                System.out.println(
+                        "Warning probabilty greater > 1 in "
+                        + Generic_BigDecimal.class
+                        + ".randomUniformTest(Random,BigDecimal). "
+                        + "Returning true.");
+                return true;
+            }
+        }
+        BigDecimal midTestValue = new BigDecimal("0.5");
+        BigDecimal half_BigDecimal = new BigDecimal("0.5");
+        BigDecimal two_BigDe
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 6848:6866, 6867:6887.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.